builder/builder-next/control/control.go

```diff
@@ -256,9 +275,30 @@ func (c *Controller) Solve(ctx context.Context, req *controlapi.SolveRequest) (*
 		return nil, err
 	}
 	if req.Exporter != "" {
-		exp, err := w.Exporter(req.Exporter, c.opt.SessionManager)
-		if err != nil {
-			return nil, err
+		var exp exporter.Exporter
+		if c.opt.UseSnapshotter {
+			if req.Exporter == mobyexporter.Moby {
+				req.Exporter = client.ExporterImage
+			}
+
+			exp, err = w.Exporter(req.Exporter, c.opt.SessionManager)
+			if err != nil {
+				return nil, err
+			}
+
+			if req.ExporterAttrs != nil {
+				reposAndTags, err := SanitizeRepoAndTags(strings.Split(req.ExporterAttrs["name"], ","))
+				if err != nil {
+					return nil, err
+				}
+				req.ExporterAttrs["name"] = strings.Join(reposAndTags, ",")
+				req.ExporterAttrs["unpack"] = "true"
```

Comment:
Looking at this again; so here we sanitize, de-duplicate, and validate the passed name(s). The results of that are written back to `req.ExporterAttrs["name"]`.

But that is passed on to `exporter.Resolve()` below. Looking at that function for the `imageExporter` (in `builder-next/exporter/containerimage`);

```go
func (e *imageExporter) Resolve(ctx context.Context, opt map[string]string) (exporter.ExporterInstance, error) {
	i := &imageExporterInstance{
		imageExporter: e,
		buildInfo:     true,
	}
	for k, v := range opt {
		switch k {
		case keyImageName:
			for _, v := range strings.Split(v, ",") {
				ref, err := distref.ParseNormalizedNamed(v)
				if err != nil {
					return nil, err
				}
				i.targetNames = append(i.targetNames, ref)
			}
```

I wonder if we're doing the same thing _twice_ now (but for the Exporter, we actually seem to be missing;

- de-duplicating
- checking if it's a `reference.Digested` (digest-only, no name)


Combined with my previous comment above, I'm now wondering; does the `Controller` need to be aware of snapshotter vs non-snapshotter, or is it enough that the `Worker` /  `WorkerController` provides the correct snapshotter to use ?

Reply:
The `builder-next/exporter/containerimage` is the exporter for moby actually so these two pieces of code don't have anything in common.

Reply:
> The `builder-next/exporter/containerimage` is the exporter for moby actually so these two pieces of code don't have anything in common.

I think my head is slightly spinning again; the `exporter/containerimage` is for graphdrivers,  correct?

```go
if b.useSnapshotter {
	exporterName = client.ExporterImage
} else {
	exporterName = mobyexporter.Moby
}
```

But it looks like we use "moby" for the exact opposite in `ContainerdWorker` ?

```go
// Exporter returns exporter by name
func (w *ContainerdWorker) Exporter(name string, sm *session.Manager) (exporter.Exporter, error) {
	switch name {
	case mobyexporter.Moby:
		return w.Worker.Exporter(client.ExporterImage, sm)
```

In the above, ` w.Worker.Exporter()` there is BuildKit's `worker/base.Worker`, which returns BuildKit's `exporter/containerimage.imageExporter`

Reply:
- `builder-next/exporter/containerimage` is the `"moby"` exporter, created [here]()
- `client.ExporterImage` is buildkit's exporter that exports things to containerd.

In `ContainerdWorker` we switch to buildkit's (containerd) image exporter if the client asks for the `"moby"` exporter.

To remove confusion we could rename the `builder-next/exporter/containerimage` to `builder-next/exporter/moby`?

Reply:
> But it looks like we use "moby" for the exact opposite in `ContainerdWorker` ?
> 
> ```go
> // Exporter returns exporter by name
> func (w *ContainerdWorker) Exporter(name string, sm *session.Manager) (exporter.Exporter, error) {
> 	switch name {
> 	case mobyexporter.Moby:
> 		return w.Worker.Exporter(client.ExporterImage, sm)
> ```

This code says: when the client asks for the moby exporter, give it this one, we do this so that older clis still work with our code.

Reply:
But in the case we have snapshotters enabled, wouldn't the `worker` already be `ContainerdWorker`? In that case;

-  `(ContainerdWorker).Exporter()` already maps `moby` -> `image`
- But the `controller` _also_ makes this mapping

Reply:
Indeed, only the worker on is needed...

Reply:
👍 removed